zipkin-common/src/test/scala/com/twitter/zipkin/storage/DependencyStoreSpec.scala

```diff
@@ -189,6 +189,93 @@ abstract class DependencyStoreSpec extends JUnitSuite with Matchers {
   }
 
   /**
+   * This test confirms that the span store can process trace with intermediate
+   * spans like the below properly.
+   *
+   *   span1: SR SS
+   *     span2: intermediate call
+   *       span3: CS SR SS CR: Dependency 1
+   */
+  @Test def getDependencies_intermediateSpans() = {
+    val trace = ApplyTimestampAndDuration(List(
+      Span(20L, "get", 20L,
+        annotations = List(
+          Annotation(today * 1000, Constants.ServerRecv, Some(zipkinWeb)),
+          Annotation((today + 350) * 1000, Constants.ServerSend, Some(zipkinWeb)))),
+      Span(20L, "call", 21L, Some(20L),
+        binaryAnnotations = List(
+          BinaryAnnotation(Constants.LocalComponent, "depth2", Some(zipkinWeb)))),
+      Span(20L, "get", 22L, Some(21L),
+        annotations = List(
+          Annotation((today + 50) * 1000, Constants.ClientSend, Some(zipkinWeb)),
+          Annotation((today + 100) * 1000, Constants.ServerRecv, Some(zipkinQuery)),
+          Annotation((today + 250) * 1000, Constants.ServerSend, Some(zipkinQuery)),
+          Annotation((today + 300) * 1000, Constants.ClientRecv, Some(zipkinWeb)))),
+      Span(20L, "call", 23L, Some(22L),
+        binaryAnnotations = List(
+          BinaryAnnotation(Constants.LocalComponent, "depth4", Some(zipkinQuery)))),
+      Span(20L, "call", 24L, Some(23L),
+        binaryAnnotations = List(
+          BinaryAnnotation(Constants.LocalComponent, "depth5", Some(zipkinQuery)))),
+      Span(20L, "get", 25L, Some(24L),
+        annotations = List(
+          Annotation((today + 150) * 1000, Constants.ClientSend, Some(zipkinQuery)),
+          Annotation((today + 200) * 1000, Constants.ClientRecv, Some(zipkinQuery))),
+      binaryAnnotations = List(
+        BinaryAnnotation(Constants.ClientAddr, true, Some(zipkinQuery)),
+        BinaryAnnotation(Constants.ServerAddr, true, Some(zipkinJdbc))))
+    ))
+
+    processDependencies(trace)
+    result(store.getDependencies(today + 1000)) should contain theSameElementsAs(dep.links)
+  }
+
+  /**
+   * This test confirms that the span store can detect dependency indicated by
+   * SERVER_ADDR and CLIENT_ADDR.
+   * In some cases an RPC call is made where one of the two services is not instrumented.
+   * However, if the other service is able to emit "sa" or "ca" annotation with a service
+   * name, the link can still be constructed.
+   *
+   *   span1: CA SR SS: Dependency 1 by a not-instrumented client
+   *     span2: intermediate call
+   *       span3: CS CR SA: Dependency 2 to a not-instrumented server
+   */
+  @Test def getDependencies_notInstrumentedClientAndServer() = {
```

Comment:
This change is linked to https://github.com/openzipkin/zipkin-dependencies-spark/pull/18